### DISCUSSION TITLE
Umail micropython and attachments missing content

### DISCUSSION THREAD

**Discussion by robertgall** [micropython/micropython#9933] [Using MicroPython] [answered]
Hi,

Using [https://github.com/shawwwn/uMail](url) to send emails with attachments.

Having issues with emailing csv/text type attachments that contain a : within the file. 
In particulat a line of text that contains a : be ommitted in attachment.
Repalcing the colon with a undescore and line of text will be included.
example
2022-04-21,09:53,Temp0,Temp1,Temp2,Temp3,Temp4,Temp5,Temp6,Temp7,Temp8,temp9,IP 
2022-04-21,09_53,Temp0,Temp1,Temp2,Temp3,Temp4,Temp5,Temp6,Temp7,Temp8,temp9,IP
First line fails to be included in the attachment, replacing the : with an _ line appers in attachment.

Using a gmail client to both send and receive the email.
[https://github.com/shawwwn/uMail](url)

example code:
smtp = umail.SMTP('smtp.gmail.com', 587, username='x', password='xx')
smtp.to('xxx')
smtp.write('Subject:datafile\n')
smtp.write('Content-Type: text/csv; name=sensor.csv\n')
smtp.write('Content-Disposition: attachment; filename=sensor.csv\n')
with open('sensor.csv', 'r') as infile:
    content=infile.read()
smtp.write(content)
smtp.write('\n')
smtp.send()
smtp.quit()

**Comment by davefes** [comment]
Hmmm, here are the first two lines out of my daily .csv file:

> Date_time,26V5 batt,PV curr,Load curr,Daily WH in,Daily WH out,SOC,Batt temp,Error,
00:01,26.44,0.0,2.5,0.0,1.1,84,24.3,0,
... and I use Gmail as well.

davef

**Comment by davefes** [comment]
My function maybe has a few hints in it.

```py
def send_csv(date):
    file_size = uos.stat('datalog.csv')[6]
    chunk_size = 1024
    counter = 0
    content = 'there is no content' #  if reading ramdisk fails this keeps things alive


    smtp = umail.SMTP('smtp.gmail.com', 587)
    utime.sleep(1)

    while True:
        counter +=1

        try:
#            test1 = smtp.login('davef@gmail.com', 'xyz')
            test1 = smtp.login('davef@gmail.com', 'abcd')

            result = str(test1).find('Accepted')

            if (result != -1):
                print('login OK')
                counter = 0
                break
        except Exception:
         #  try to capture the failure
            try:
                with open('errors.txt', 'a') as outfile:
                    outfile.write('smtp.login did NOT work')
            except OSError:
                pass

        if (counter == 5):
            counter = 0

         #  generate a wdt
            utime.sleep(80)

    counter = 0 #  reset

    wdt_feed()
    
    utime.sleep(1)
    smtp.to('davef@gmail.com')
    utime.sleep(1)
    smtp.write('From: davef@gmail.com\n')
    smtp.write('To: davef@gmail.com\n')
    smtp.write('Subject: Today\'s ' + config.device + 'datalog\n')
    smtp.write('MIME-Version: 1.0\n')
    smtp.write('Content-type: multipart/mixed; boundary=12345678900987654321\n')

    smtp.write('--12345678900987654321\n')
    smtp.write('Content-Type: text/plain; charset=utf-8\n')
    smtp.write('see attachment\n')
    smtp.write('--12345678900987654321\n')

 #  build today's filename into Content-Type and Content-Disposition
    csv_file = config.device + date + '.csv\n'

    content_type = 'Content-Type: text/csv; name=' + csv_file
    content_disposition = 'Content-Disposition: attachment; filename=' + csv_file 
    smtp.write(content_type)
    smtp.write(content_disposition)

    wdt_feed()
    
 #  send data in 1024 chunks
    try:
        with open('datalog.csv', 'rb') as file:
            for chunk in range(0, file_size, chunk_size):
                this_chunk_size = min(chunk_size, file_size-chunk)
             #  handles the last dangling chunk that would be smaller than the full 1024 bytes
                smtp.write(file.read(this_chunk_size))
    except OSError as error:
        try:
            with open('errors.txt', 'a') as outfile:
                outfile.write('error = ' + str(error) + '\n')
                outfile.write('did not handle datalog.csv properly' + '\n')
        except OSError:
            pass

    smtp.send() # does the proper ending
    smtp.quit()

    print('log has been sent')
```

**Comment by karfas** [comment, marked as answer]
The RFCs (see e.g. wikipedia on SMTP) states that all lines need to end with CRLF ("\r\n" in python). 
This _might_ be handled by the umail package, but who knows without looking ?

I think the real error here is that you don't separate the mail headers from the content. 
In SMTP needs to be an empty line after you send the content-type, e.g. 
```
smtp.write('Content-Disposition: attachment; filename=sensor.csv\r\n')
smtp.write('\r\n')
with open('sensor.csv', 'r') as infile:
    content=infile.read()
``` 
My guess: everything in front of the colon `:` gets interpreted as some (unknown) header when you don't tell the package that the mail headers are complete.

**Reply by robertgall** [reply]
Thanks Karfas (and Dave),

Not sure if still need to add the carriage return and probably good practice to do, but seems like adding the additional \n did the trick!

smtp.write('Content-Disposition: attachment; filename=sensor.csv\n\n')
with open('sensor.csv', 'r') as infile:

Rob.
